Sources/Markdown/Interpretive Nodes/Aside.swift

```diff
@@ -170,51 +180,54 @@ public struct Aside {
             // Otherwise, default to a note aside.
             self.kind = .note
             self.content = Array(blockQuote.blockChildren)
             return
         }
-        self.kind = Kind(rawValue: String(initialText.string[..<firstColonIndex]))!
 
-        // Trim off the aside tag prefix.
-        let trimmedText = initialText.string[initialText.string.index(after: firstColonIndex)...].drop {
-            $0 == " " || $0 == "\t"
+        self.kind = Kind(rawValue: kindTag.tag)!
+        self.content = Array(kindTag.newBlockQuote.blockChildren)
+    }
+
+    /// Create an aside from a block quote if it contains a kind tag that matches the given requirement.
+    public init?(_ blockQuote: BlockQuote, tagRequirement: TagRequirement = .tagNotRequired) {
+        guard tagRequirement != .tagNotRequired else {
+            self = .init(blockQuote)
+            return
         }
-        initialText.string = String(trimmedText)
 
-        let newBlockQuote = initialText.parent!.parent! as! BlockQuote
-        self.content = Array(newBlockQuote.blockChildren)
+        guard let kindTag = blockQuote.parseAsideTag(tagRequirement: tagRequirement) else {
+            return nil
+        }
+
+        self.kind = Kind(rawValue: kindTag.tag)!
+        self.content = Array(kindTag.newBlockQuote.blockChildren)
     }
 }
 
 extension BlockQuote {
-    /// Conservatively checks the text of this block quote to see whether it can be parsed as an aside.
-    ///
-    /// Whereas ``Aside/init(_:)`` will use all the text before the first colon in the first line,
-    /// or else return an ``Aside`` with a ``Aside/Kind`` of ``Aside/Kind/note``,
-    /// this function will allow parsers to only parse an aside if there is a single-word aside
-    /// marker in the first line, and otherwise fall back to a plain ``BlockQuote``.
-    func isAside() -> Bool {
-        guard let initialText = self.child(through: [
+    func parseAsideTag(tagRequirement: Aside.TagRequirement) -> (tag: String, newBlockQuote: BlockQuote)? {
+        guard var initialText = self.child(through: [
             (0, Paragraph.self),
             (0, Text.self),
-        ]) as? Text else {
-            return false
+        ]) as? Text, initialText.string.contains(":") else {
+            return nil
         }
 
-        for character in initialText.string {
-            switch character {
-                case ":": return true  // encountered ":" before " "
-                case " ": return false // encountered " " before ":"
-                default:  continue
-            }
+        let firstColonIndex = initialText.string.firstIndex(of: ":")!
+        let kindTag = initialText.string[..<firstColonIndex]
+        let trimmedText = initialText.string[initialText.string.index(after: firstColonIndex)...].drop {
+            $0 == " " || $0 == "\t"
+        }
+        initialText.string = String(trimmedText)
+        let newBlockQuote = initialText.parent!.parent! as! BlockQuote
```

Comment:
Following up on [this comment][1]: 
If I add these assertions to the code and run the tests, then I hit the 2nd assertion (and 3rd assertion if I comment out the 2nd).

[1]: https://github.com/swiftlang/swift-markdown/pull/196/files#r1758423514


```diff
+ assert(initialText.range?.lowerBound.source == self.range?.lowerBound.source, "Parsing didn't lose the original source information")
  initialText.string = String(trimmedText)
+ assert(initialText.range?.lowerBound.source == self.range?.lowerBound.source, "Parsing didn't lose the original source information")
  let newBlockQuote = initialText.parent!.parent! as! BlockQuote
+ assert(newBlockQuote.range?.lowerBound.source == self.range?.lowerBound.source, "Parsing didn't lose the original source information")
```

Reply:
It looks like the `initialText.string = String(trimmedText)` line is clearing the range for the text node, which in turn is clearing the range for the block quote as a whole. This bug was present in the old version of the code for any Asides that are parsed, since this code was adapted from the existing Aside initializer, but we can fix it for the future.

Reply:
It involves doing a lot of tree copies since the source range is part of the `RawMarkupHeader` type. I believe i have a fix written on my machine, i just need to make sure it works as expected.